docs/exec_model.md

```diff
@@ -0,0 +1,154 @@
+Worker has its own task queues and execution engine. Each worker has its own JS
+context and Workers are executed concurrently (e.g. one OS thread per Worker).
+
+Communication between Workers is done via message passing. Using a shared memory
+was possible through
+[SharedBufferArray](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/SharedArrayBuffer)
+but it has been disabled for security reasons.
+
+**Wasm threads**
+
+There is a
+[proposal](https://github.com/WebAssembly/threads/blob/master/proposals/threads/Overview.md)
+to add support for threads in WebAssembly reusing Worker and SharedBufferArray
+but it isn't implemented yet.
+
+## Haskell execution model
+
+Haskell threads are represented by a Thread State Object (TSO). Each TSO has its
+own Stack.
+
+In the native RTS, TSOs are executed by Capabilities (virtual CPUs). Each
+Capability has a run queue of TSOs and also maintains lists of TSOs blocked for
+different reasons. Capabilities are executed by Tasks, where a Task is a wrapper
+for an OS thread. Task specific data are stored in thread-local storage (TLS). A
+Task owns a single Capability at once.
+
+The Haskell Heap is shared between all the Haskell threads. TSOs themselves
+are objects in the heap.
+
+**Scheduling**
+
+TSOs can be preempted when they call into the runtime system. It can be:
+
+* when they
+  [yield](https://www.stackage.org/haddock/lts-14.5/base-4.12.0.0/Control-Concurrent.html#v:yield) (cooperative scheduling)
+* when they call into blocking functions (``takeMVar``, etc.)
+* when they call into FFI functions
+* when they don't have enough heap space to allocate
+
+The runtime system may force a TSO preemption by faking a reduction of the
+heap space to 0. The preemption only happens when the TSO tries to allocate
+though (hopefully often enough).
+
+## Executing Haskell code with JS engines
+
+We somehow need to map Haskell execution model on JavaScript one.
+
+### Step 0: no multi-threading
+
+The global shared memory is a WebAssembly.Memory object (see
+[memory](memory.md)). Currently it can't be shared between several JavaScript
+contexts.
+
+Currently Asterius [doesn't support
+multi-threading](https://github.com/tweag/asterius/issues/268). A single TSO is
+executed synchronously until it finishes, blocking any other macro-task.
+
+In practice: ``rts_eval*`` functions in ``rts.Exports.mjs`` create a new thread
+(using ``createThread`` function in Asterius.Builtins) and execute it
+synchronously with ``scheduleWaitThread`` defined in Asterius.Builtins.
+
+### Step 1: cooperative multi-threading
+
+We could introduce a run queue of TSOs. Each time a TSO calls back into the
+scheduler we can schedule another TSO.
+
+We can postpone the execution of the next TSO with
+``setTimeout(executeNextTSO,0)`` in order to let the JS engine execute other
+macro-tasks in between. E.g. JSFFI using ``setTimeout``, browser events, etc.
+
+### Step 2: cooperative multi-threading + Worker
```

Comment:
Even when the Spectre mitigation is complete in all major browsers, we should be careful about `Worker` usage and not rely on it by default; it's still heavyweight imho. Running everything in the main thread is not going to block execution of other JavaScript functions noticeably, because our runtime is async by nature and has a lot of chances of yielding control to other functions (e.g. ongoing async calls)